QA: strict types

arueckauer opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | yes

### Description

This PR provides strict typing as described in #2 .

- Add strict type declaration
- Add return type declaration
- Add argument type declaration (where applicable)
- Fix type errors
- Remove tests which are made redundant by type declaration
- Fix second parameter type in calls of `var_export`

arueckauer committed:
Add strict type declaration

arueckauer committed:
Add class, function and constant imports

arueckauer committed:
Declare return types

arueckauer committed:
Remove unnecessary fully qualifier

arueckauer committed:
Update return type declaration

arueckauer committed:
Declare argument types

arueckauer committed:
Fix type errors for Client

arueckauer committed:
Restore assertions
They add a valid test case, unable to find out the reason why they were disabled.

arueckauer committed:
Fix type errors in Error
Add initial value for
Update tests
Remove tests which are made redundant by type declaration

arueckauer committed:
Fix type errors in Request
Allow non-string values for parameter key (for cases where params are provided by position)
Drop string cast of  (JSON-RPC 1.0 allows any type)
Update tests Fix second parameter type in calls of var_export

arueckauer committed:
Fix type errors in Response
Remove test data which is made redundant by type declaration

arueckauer committed:
Fix type error caused by passing a numeric value of  to ucfirst method
Error was introduced by strict type declaration

arueckauer committed:
Fix type errors in Server
Add missing return statement
Fix second parameter type in calls of var_export
Use ::class constant as FQCN reference

arueckauer committed:
Fix type errors in Smd
Update tests

arueckauer committed:
Fix type errors in Smd\Service
Update tests
Remove tests which are made redundant by type declaration
Fix second parameter type in calls of var_export

arueckauer committed:
Fix CS (1 space between comma and argument)

arueckauer committed:
Change test method name to reflect what is being tested (expecting no exception being thrown).

arueckauer committed:
Ignore exceeding of line length in CS for now
Added return type declarations caused this

michalbundyra commented on src/Error.php:
```diff
@@ -38,14 +40,14 @@ class Error
      * @param  mixed $data
      */
-    public function __construct($message = null, $code = self::ERROR_OTHER, $data = null)
+    public function __construct(?string $message = '', int $code = self::ERROR_OTHER, $data = null)
```
Probably `$message` shouldn't be nullable.

arueckauer replied:
Agreed. Change will be pushed shortly.

michalbundyra commented on src/Exception/ExceptionInterface.php:
```diff
@@ -6,6 +6,8 @@
+
 namespace Laminas\Json\Server\Exception;
 
 interface ExceptionInterface
```
We want to add `extends Throwable` here.

arueckauer replied:
Agreed. I will open an issue as reminder and provide changes separately.

michalbundyra commented on src/Error.php:
```diff
@@ -97,13 +89,9 @@ public function getCode()
      * @return self
      */
-    public function setMessage($message)
+    public function setMessage(string $message) : self
```
We tend to drop fluent interfaces. It might be out of scope of that PR, but something what should be done noted and done for the next major.

arueckauer replied:
Agreed. I will open an issue as reminder and provide changes separately.

michalbundyra commented on src/Error.php:
```diff
@@ -158,7 +146,7 @@ public function toArray()
-    public function toJson()
+    public function toJson() : string
     {
         return Json::encode($this->toArray());
```
This should be confirmed, and also not for that PR: If I remember correctly we are getting rid of Json::encode / Json::decode in favour of requiring ext-json in composer and json_* functions.

About strict types here: `$data` cannot be completely mixed, must be 'serializable'.

arueckauer replied:
JSON: Alright, I will open an issue as reminder.

`$data`: I will have to have a closer look to get a better understanding of the context and usage. I will open an issue as reminder.

michalbundyra commented on src/Request.php:
```diff
@@ -235,7 +255,7 @@ public function getId()
-    public function setVersion($version)
+    public function setVersion(string $version) : self
     {
         if ('2.0' == $version) {
```
String comparison here as we have now param type declaration.

For another PR: we tend to avoid yoda conditions.

arueckauer replied:
Strict comparison: Change will be pushed shortly.

Yoda conditions: Noted. I will open an issue.

michalbundyra commented on src/Response.php:
```diff
@@ -140,7 +151,7 @@ public function getResult()
      * @return self
      */
-    public function setError(Error $error = null)
+    public function setError(Error $error = null) : self
```
If I remember correctly it will be enforced for LCS 2.0 - typehint for nullable parameters should be always with `?`. Can be addressed later with other CS changes.

(Also just to note: with LCS 2.0 we go with PSR-12 return type hint formatting, so no space before `:`).

arueckauer replied:
Nullable type declaration: I tend to add it, must have missed this one. Will provide the change shortly.

Same for no space before return type declaration colon.

michalbundyra commented on src/Response.php:
```diff
@@ -194,9 +205,8 @@ public function getId()
+    public function setVersion(?string $version) : self
     {
-        $version = (string) $version;
         if ('2.0' == $version) {
```
strict comparison here

(also no yoda, but this can be addressed later).

arueckauer replied:
Will push strict comparison shortly.

michalbundyra commented on src/Response.php:
```diff
@@ -194,9 +205,8 @@ public function getId()
      * @return self
      */
-    public function setVersion($version)
+    public function setVersion(?string $version) : self
```
Here we have inconsistent param type declaration with PHPDocs.
With LCS 2.0 we are not duplicating PHPDocs if nothing is changes. So this can be addressed later.

arueckauer replied:
Understood.

michalbundyra commented on src/Response.php:
```diff
@@ -277,7 +287,7 @@ public function setServiceMap($serviceMap)
      * @return Smd|null
      */
-    public function getServiceMap()
+    public function getServiceMap() : ?Smd
```
This is a bit confusing - in theory `->serviceMap` can be `null`, but only when it is not initialised.
We would need to decide how we are doing it. I think for that case I would prefer drop nullable type here.
With PHP 7.4 if we add property type declaration (`protected Smd $serviceMap;`) we get error when we try get value of uninitialised property.

arueckauer replied:
As in the current implementation the `$serviceMap` property is optional, nullable looks fine to me. Removing the nullable will break `Response\Http::sendHeaders()`.
This seems to be an issue that needs to addressed with update to PHP 7.4.

michalbundyra replied:
If this is optional I think setter should allow nullable values. Thoughts?

arueckauer replied:
Agreed.

michalbundyra commented on src/Server.php:
```diff
@@ -77,7 +101,7 @@ class Server extends AbstractServer
      * @throws Exception\InvalidArgumentException If function invalid or not callable.
      */
-    public function addFunction($function, $namespace = '')
+    public function addFunction($function, $namespace = '') : Server
```
Missing param types here.

Is it because we extend abstract?
Maybe in that case will be worth also update laminas-server first?

arueckauer replied:
Correct, method signature is defined by `Server` interface of `laminas/laminas-server`.
I am willing to help upgrade that package, too. Since I will be on a leave and unavailable in February and March, I would like to complete this package first - if possible.

michalbundyra commented on src/Server.php:
```diff
@@ -549,17 +573,17 @@ protected function getSmdMethods()
      * Internal method for handling request.
      *
-     * @return void
+     * @return Error
```
Inconsistent with declared RTH on the method.
But as noted before, it should be removed completely as no additional information provided.

arueckauer replied:
Fixed PHPDocs for now. Removal of unnecessary PHPDocs will be addressed separately.

michalbundyra commented on src/Smd.php:
```diff
@@ -122,9 +132,9 @@ public function setOptions(array $options)
      * @param  string $transport
      * @return self
-     * @throws Exception\InvalidArgumentException
+      * @throws InvalidArgumentException
```
Additional space before `*`.

arueckauer replied:
👍🏻

michalbundyra commented on src/Smd.php:
```diff
@@ -250,18 +260,18 @@ public function getId()
+    public function setDescription(string $description) : self
     {
-        $this->description = (string) $description;
-        return $this->description;
```
I think this may be considered as a bug in the current version.
We would need a bugfix for it (PR into master).

arueckauer replied:
Noted. I will open an issue.

michalbundyra commented on src/Smd.php:
```diff
@@ -322,7 +332,7 @@ public function addService($service)
      * @return self
      */
-    public function addServices(array $services)
+    public function addServices(array $services) : self
```
In general for cases like that we want to use something more generic than `array`, probably `iterable`.

arueckauer replied:
Good point. I would like to open an issue for that to handle it separately. Are you okay with that?

michalbundyra commented on src/Smd.php:
```diff
@@ -293,10 +303,10 @@ public function isDojoCompatible()
+     * @throws InvalidArgumentException
      */
-    public function addService($service)
+    public function addService($service) : self
```
We need to consider here just allowing `Smd\Service` here and then we can drop checks inside. Thoughts?

arueckauer replied:
I agree, this would make the purpose of `$service` more precise. I will open an issue.

michalbundyra commented on src/Smd/Service.php:
```diff
@@ -26,9 +44,9 @@ class Service
+    protected $name = '';
     protected $return;
-    protected $target;
+    protected $target = '';
```
we need php docs for each property with type definition.

arueckauer replied:
As I understand it, this is a block of properties that are defined in the SMD specs. Do you know what the specials chars (`/**#@+` and ` /**#@-*/`) in the comment are for? Are they a directive to skip CS checks/fixes?

michalbundyra replied:
I knew what it was.... But I forgot... :(
It was some notation for the IDE (vim?) but we don't need it anymore and PHPDocs for each property is needed.

michalbundyra commented on src/Smd/Service.php:
```diff
@@ -26,9 +44,9 @@ class Service
      */
     protected $envelope  = Smd::ENV_JSONRPC_1;
-    protected $name;
+    protected $name = '';
```
Not sure if I like this change. As I mentioned before - with PHP 7.4 it will be a bit easier. I think `null` for no initialised is fine now (looking at line 142)

arueckauer replied:
Removing empty string initialization means, that the return type of the getter is nullable. Would that not be confusing?

Also the [specs](https://github.com/dojo/docs/blob/master/dojox/rpc/smd.rst) state for `name`
> This should be unique for each method

From my understanding a nullable type is optional **and** not unique. Personally, I would prefer an empty string initialization.

Agreed on line 142, this should be a strict comparison.

michalbundyra replied:
Ok, I think it's fine, because we have check in the constructor and providing empty name should be also invalid.

michalbundyra reviewed (commented):
@arueckauer Thanks for the PR. Please see my comments. Some of them are just notes for later. In general it looks good. Maybe on some cases we can change to use just one specific type instead of allowing multiple options.

As noted in some comments:
- we are dropping fluent interfaces
- we are using no space before colon in RTH declaration (PSR-12)
- `null` is fine for uninitialised properties (imo) and then we need decide what with RTH on getter

arueckauer committed:
Remove space before return type declaration colon
In preparation for LCS 2.0

arueckauer committed:
Remove nullable argument type for

arueckauer committed:
Use strict comparison
... has now a declared type of string

arueckauer committed:
Update type declaration to be nullable for nullable argument
In preparation for LCS 2.0

arueckauer committed:
Use strict comparison
... has now a declared type of string

arueckauer committed:
Update PHPDocs to match return type declaration

arueckauer committed:
Remove additional space in PHPDocs line

arueckauer committed:
Use strict comparison
... has now a declared type of string

arueckauer committed:
Add argument and return type declarations to service class example

arueckauer commented:
👋🏻 and thanks for the review. That's great input! Most comments are fine, for a few I had a question.

Since this PR is about adding strict typing to the component. I would like to address changes not related to it separately. Would you be okay with me opening issues for the following list? They would be addressed in `dev-4.0` (`master` respectively for bugfixes).

- Drop fluent setters
- Avoid yoda conditions
- Drop `Json::encode` & `Json::decode` in favor of native `json_encode` and `json_decode` (add ext-json to composer.json)
- Drop PHPDocs which do not add valuable information (such as duplicating type declarations)
- `ExceptionInterface` should extend `Throwable`
- Apply `laminas/laminas-coding-standard` 2.0
- Bugfix `Smd::getDescription()` returns `$description` instead of `self`
- Handle `array` arguments more generic, probably `iterable` (e.g. `Smd::addServices(array $services)`)
- `Smd::addService(Smd\Service $service)` drop array support and argument type check [BC]
- JSON-RPC SMD: Use argument type declarations as param types (deprecating type hints in PHPDocs)

EDIT:
Added _JSON-RPC SMD: Use argument type declarations as param types (deprecating type hints in PHPDocs)_ to list of changes.

michalbundyra commented:
@arueckauer Maybe add these as checkboxes to #2 (edit the description) and then provide separate PRs to dev-4.0. What do you think?

EDIT:

> Bugfix Smd::getDescription() returns $description instead of self

PR to master please.

arueckauer committed:
Allow null value in setter for optional
Update test (previously failed due to null being an unsupported JSON-RPC request param type)

arueckauer committed:
Add missing PHPDocs

arueckauer commented:
Agreed, I added the list.

There is one unresolved conversation about initial value of `Smd\Service::$name`. Would love your input on that.

arueckauer committed:
Add strict comparison argument to use type comparison in in_array usage

arueckauer committed:
Use strict comparison

Xerkus commented on src/Cache.php:
```diff
@@ -56,10 +70,9 @@ public static function saveSmd($filename, Server $server)
      * @return string|false
      */
-    public static function getSmd($filename)
+    public static function getSmd(string $filename)
```
Design:
Return should be changed to `?string` As you get a result or nothing.

arueckauer replied:
Noted.

Xerkus commented on src/Cache.php:
```diff
@@ -24,10 +38,10 @@ class Cache extends ServerCache
      * @return bool
      */
-    public static function saveSmd($filename, Server $server)
+    public static function saveSmd(string $filename, Server $server): bool
```
Design:
return should be `void`. Regardless of whether caching attempt was successful just carry on.

arueckauer replied:
Noted.

Xerkus commented on src/Cache.php:
```diff
@@ -86,9 +99,9 @@ public static function getSmd($filename)
      * @return bool
      */
-    public static function deleteSmd($filename)
+    public static function deleteSmd(string $filename): bool
```
Design:
Operation result is irrelevant for the consumer. Return nothing. `:void`

arueckauer replied:
Noted.

michalbundyra commented:
@arueckauer Also for strict types we should use `assertSame` instead of `assertEquals` (the same as `===` instead of `==`).

Xerkus commented on src/Client.php:
```diff
@@ -54,7 +56,7 @@ class Client implements ServerClient
      * @param HttpClient $httpClient HTTP Client to use for requests.
      */
-    public function __construct($server, HttpClient $httpClient = null)
+    public function __construct(string $server, HttpClient $httpClient = null)
```
QA: rename  `$server` to `$serverAddress` to better reflect its meaning

arueckauer replied:
Noted.

Xerkus commented on src/Request.php:
```diff
@@ -72,17 +92,17 @@ class Request
      * @param  array $options
      * @return self
```
change `@return self` to `@return $this` anywhere fluid interface is used. It is a special phpdoc return type for cases like here. Apply across the repository.

Xerkus replied:
There will be another QA pass removing fluid interfaces, ignore this.

arueckauer replied:
Noted.

Xerkus commented on src/Server.php:
```diff
@@ -309,17 +333,17 @@ public function getReturnResponse()
+    public function __call(string $method, array $args)
     {
         if (! preg_match('/^(set|get)/', $method, $matches)) {
             return;
```
Design:
`return null;` should be used when method is not `: void`. A change to be done when static analysis tool introduced.,

arueckauer replied:
Noted.

Xerkus commented on src/Server.php:
```diff
@@ -549,17 +573,17 @@ protected function getSmdMethods()
         }
 
-        if (! $request->isMethodError() && null === $request->getMethod()) {
+        if (! $request->isMethodError() && '' === $request->getMethod()) {
```
`empty()` would be better here and  in similar cases as it makes code more readable: reading without IDE, eg during code review, I do not really know method signature. That makes it harder to verify validity on a passing glance.
 
It is different to checking for `null` as later means no value at all and phpstan will complain about comparing it to non-nullable type
/cc @michalbundyra

Xerkus replied:
QA: Method error check block from below can be merged with this block:
`if ($request->isMethodError()  || empty($request->getMethod())) {`
Alternatively blocks can be swapped and condition simplified to check only for empty method.

arueckauer replied:
@Xerkus Just so I get it right: Instead of an empty string strict comparison `$variable === ''`, `empty` construct should be used `empty($variable)` in general?

Xerkus replied:
I will not be asking you to change it to `empty()`.

There is a bit of a conundrum here:
- `empty()` does some special form of type juggling. That is bad. Eg `'0'` is considered empty.
- `=== ''` condition validity is not immediately clear during code review, in this case without checking return type declaration.
- value type not enforced by type declaration or type check must be treated as uncertain and potentially nullable.
- static analysis will spot `null === $request->getMethod()` as always false and will complain. But it will not notice `'' === $request->getMethod()` not covering `null`
- What makes `empty()` more clear is that it covers `null === $request->getMethod() || '' === $request->getMethod()`. It works regardless of value nullability and as such does not open a way for hard to spot logic error. 
- `empty()` treating certain values as empty opens a way for hard to spot logic errors when those values can appear among valid values.

arueckauer replied:
ACK

Xerkus commented on test/TestAsset/TestIteratorAggregate.php:
```diff
@@ -21,7 +23,7 @@ class TestIteratorAggregate implements IteratorAggregate
     ];
 
-    public function getIterator()
+    public function getIterator(): ArrayIterator
```
use  special `: iterable` type

Xerkus replied:
Sorry, return type is `Traversable`, not `iterable` : https://www.php.net/manual/en/iteratoraggregate.getiterator.php

arueckauer replied:
Added.

Xerkus commented on src/Server.php:
```diff
@@ -77,7 +101,7 @@ class Server extends AbstractServer
      * @throws Exception\InvalidArgumentException If function invalid or not callable.
      */
-    public function addFunction($function, $namespace = '')
+    public function addFunction($function, $namespace = ''): Server
```
```suggestion
    public function addFunction(callable $function, string $namespace = ''): Server
```

arueckauer replied:
Suggested declaration is not compatible with super.

Xerkus commented on src/Server.php:
```diff
@@ -77,7 +101,7 @@ class Server extends AbstractServer
-    public function addFunction($function, $namespace = '')
+    public function addFunction($function, $namespace = ''): Server
     {
         if (! is_callable($function)) {
```
Drop as replaced by typehint

arueckauer replied:
I don't follow. What should be dropped?

michalbundyra replied:
@Xerkus thought we can add param type hint `callable` and drop the if statement here, but we can't because it comes from parent.

I am suggesting updating laminas-server before, but this is a lot of job as well...

arueckauer replied:
Agreed. These are changes that I would like to see.

Can we complete this PR, then update `laminas-server` and afterwards update usage in `laminas-json-server`?

arueckauer replied:
I made a note about the update of `laminas-server` and thus will resolve this conversation.

Xerkus commented on src/Server.php:
```diff
@@ -128,7 +152,7 @@ public function addFunction($function, $namespace = '')
      * @return Server
      */
-    public function setClass($class, $namespace = '', $argv = null)
+    public function setClass($class, $namespace = '', $argv = null): Server
```
Any reason those parameter typehints are not added?

Xerkus replied:
```suggestion
    public function setClass(string $class, string $namespace = '', $argv = null): Server
```

arueckauer replied:
Argument type declarations are not compatible with super.

Xerkus commented on src/Server.php:
```diff
@@ -200,7 +224,7 @@ public function handle($request = false)
      * @return void
      */
-    public function loadFunctions($definition)
+    public function loadFunctions($definition): void
```
This looks like a candidate for `iterable`

arueckauer replied:
Again, argument type declaration is not compatible with super.

Xerkus commented on src/Server.php:
```diff
@@ -286,7 +310,7 @@ public function getResponse()
      * @return self
      */
-    public function setReturnResponse($flag = true)
+    public function setReturnResponse($flag = true): self
```
```suggestion
    public function setReturnResponse(bool $flag = true): self
```

arueckauer replied:
Suggested declaration is not compatible with super.

Xerkus commented on src/Server.php:
```diff
@@ -153,7 +177,7 @@ public function setClass($class, $namespace = '', $argv = null)
      * @return Error
      */
-    public function fault($fault = null, $code = 404, $data = null)
+    public function fault($fault = null, $code = 404, $data = null): Error
```
```suggestion
    public function fault(string $fault = '', int $code = 404, $data = null): Error
```

michalbundyra replied:
@Xerkus We can't because this is implementation of abstract/interface from laminas-server.... Unless we update first laminas-server, what I've suggested.

arueckauer replied:
As @michalbundyra mentioned, argument type declaration is not compatible with super. Since I will only be available until the end of next week and then be gone for two months, I would prefer completing this PR first.

Xerkus commented on src/Server.php:
```diff
@@ -167,7 +191,7 @@ public function fault($fault = null, $code = 404, $data = null)
      * @throws Exception\InvalidArgumentException
      */
-    public function handle($request = false)
+    public function handle($request = false): ?Response
```
Not sure about the reason for false instead of null here. This should be changed at some point i think

arueckauer replied:
Noted.

Xerkus commented on src/Server.php:
```diff
@@ -286,7 +310,7 @@ public function getResponse()
-    public function setReturnResponse($flag = true)
+    public function setReturnResponse($flag = true): self
     {
         $this->returnResponse = (bool) $flag;
```
```suggestion
         $this->returnResponse = $flag;
```

Xerkus replied:
Also add method typehint

arueckauer replied:
Unfortunately this is not compatible with super.

Xerkus commented on src/Server.php:
```diff
@@ -376,7 +400,7 @@ protected function addMethodServiceMap(Method\Definition $method)
      * @return string
      */
-    protected function _fixType($type)
+    protected function _fixType($type): string
```
```suggestion
    protected function _fixType(string $type): string
```

arueckauer replied:
Unfortunately this is not compatible with super.

Xerkus commented on test/CacheTest.php:
```diff
@@ -41,37 +49,37 @@ protected function setUp() : void
     }
 
-    public function testRetrievingSmdCacheShouldReturnFalseIfCacheDoesNotExist()
+    public function testRetrievingSmdCacheShouldReturnFalseIfCacheDoesNotExist(): void
```
I don't really care if return type declared on tests. They provide no impact on code quality. I personally prefer not to have them to save me some effort when writing tests and also save extra 6 characters for lengthy verbose test names

arueckauer replied:
I get confused having different CS definitions within the same project, especially as I have test and implementation classes next to each other during development. Maybe these method names will be refactored during the coming changes.

@Xerkus Would it be okay for you to keep return type declarations or would you like them removed?

Xerkus replied:
Either way works, no established rule.  Just keep it consistent within the repo

arueckauer replied:
👍🏻

Xerkus reviewed (changes requested):
There are few additional changes needed.
Also I left couple notes about things that should be addressed in the future.

arueckauer committed:
Apply strict assertions by replacing assertEquals() through assertSame()
Update assertions where necessary to fulfill strict comparison

arueckauer committed:
Update return type of getIterator() to Traversable

michalbundyra reviewed (approved):
@arueckauer LGTM 👍

Xerkus commented on src/Response.php:
```diff
@@ -194,10 +205,9 @@ public function getId()
      * @return self
      */
-    public function setVersion($version)
+    public function setVersion(?string $version): self
```
phpdoc does not match

arueckauer replied:
Fixed.

Xerkus replied:
@arueckauer `|null` added to wrong tag

michalbundyra replied:
@Xerkus nice catch but there is no point of blocking this PR by issues like that. These will be detected later on by tools (phpstan / LCS). Just go ahead and merge it, so @arueckauer can progress on the component.

Xerkus commented on src/Response.php:
```diff
@@ -264,9 +274,9 @@ public function setArgs($args)
      *
      * @param  Smd $serviceMap
-     * @return self
+     * @return self|null
```
`null` should have been added to `@param` instead

arueckauer replied:
Fixed.

Xerkus commented on src/Smd/Service.php:
```diff
@@ -470,7 +500,7 @@ public function __toString()
-    protected function validateParamType($type, $isReturn = false)
+    protected function validateParamType(string $type, bool $isReturn = false): string
     {
         if (! is_string($type)) {
```
drop redundant check

arueckauer replied:
Fixed.

Xerkus reviewed (changes requested):
only couple minor changes left

arueckauer committed:
Fix return type in PhpDocs

arueckauer committed:
Set null as default value for  to have a clrearer not-present state for this optional parameter

arueckauer committed:
Fix falsely placed '|null'

arueckauer committed:
Drop redundant type check
